<span style="color:darkblue"> ⚠️ <b>IMPORTANTE</b></span>: <b>CORRE A CAIXA DE CÓDIGO SEGUINTE</b> antes de começares as tuas experiências!
<br><span style="color:gray">(Para isso, clica na caixa e depois carrega no botão ![image.png](./assets/play.PNG) que está na barra acima).</span>

In [ ]:
:opt no-lint
:l scripts/ensico.hs
putStrLn (['F']++['e','i','t']++"o!")

---

# Soluções dos exercícios do Trabalho de Grupo 1 (2023/24)
> **Computação - Ano 2 - Aula 4** (Aula Prática)<br>
> © 2020-2024 ENSICO. Todos os direitos reservados

## Perguntas 2 a 5
A seguinte lista numérica tem as <u>idades de todos os alunos de 6º</u> ano do AEFGA.

`[11,10,11,11,11,10,11,11,11,11,11,11,11,11,11,11,11,10,11,11,11,10,11,11,11,11,10,11,12,11,11,12,11,11,11,11,11,11,10,11,10,11,11,11,11,11,11,11,10,10,11,11,11,12,11,12,11,11,11,10,11,12,11,10,11,14,11,11,10,11,11,13,11,10,11,11,11,11,11,11,10,11,10,11,11,11,12,11,10,11,11,11,11,11,11,11,10,11,11,12,11,12,11,11,11,11,11,10,11,11,11,10,11,10,10,11,10,10,10,10,11,10,10,11,11,12,11,11,11,11,11,11,10,10,10,11,11,11,11,11,11,10,11,11,11,11,11,11,11,11,11,10,11,11,11,11,10]`

### Pergunta 2
Qual a <u>idade do aluno mais velho</u> do 6º ano?

**Solução:**

In [2]:
idades6 = [11,10,11,11,11,10,11,11,11,11,11,11,11,11,11,11,11,10,11,11,11,10,11,11,11,11,10,11,12,11,11,12,11,11,11,11,11,11,10,11,10,11,11,11,11,11,11,11,10,10,11,11,11,12,11,12,11,11,11,10,11,12,11,10,11,14,11,11,10,11,11,13,11,10,11,11,11,11,11,11,10,11,10,11,11,11,12,11,10,11,11,11,11,11,11,11,10,11,11,12,11,12,11,11,11,11,11,10,11,11,11,10,11,10,10,11,10,10,10,10,11,10,10,11,11,12,11,11,11,11,11,11,10,10,10,11,11,11,11,11,11,10,11,11,11,11,11,11,11,11,11,10,11,11,11,11,10]

In [3]:
idades6 |> maximum

14

### Pergunta 3
<u>Quantos alunos de 6º ano</u> há no AEFGA?

**Solução:**

In [4]:
idades6 |> length

157

### Pergunta 4
<u>Quantas idades distintas</u> existem no 6º ano?

<small><b>Nota:</b> "distintas" significa "diferentes umas das outras".</small>
    
**Solução:**

Começamos por tirar as idades repetidas da lista de idades, aplicando a transformação `nub`.

In [5]:
idades6 |> nub

[11,10,12,14,13]

Depois contamos quantas idades distints sobraram nessa lista, usando a transformação `length`:

In [6]:
idades6 |> nub |> length

5

### Pergunta 5
<u>Qual é a lista</u> de idades que os alunos do 6º ano vão ter <u>daqui a 5 anos</u>?

<small><b>Nota:</b> Copia o resultado dado pelo Jupyter.</small>

**Solução:**

In [7]:
idades6 |> map (+5)

[16,15,16,16,16,15,16,16,16,16,16,16,16,16,16,16,16,15,16,16,16,15,16,16,16,16,15,16,17,16,16,17,16,16,16,16,16,16,15,16,15,16,16,16,16,16,16,16,15,15,16,16,16,17,16,17,16,16,16,15,16,17,16,15,16,19,16,16,15,16,16,18,16,15,16,16,16,16,16,16,15,16,15,16,16,16,17,16,15,16,16,16,16,16,16,16,15,16,16,17,16,17,16,16,16,16,16,15,16,16,16,15,16,15,15,16,15,15,15,15,16,15,15,16,16,17,16,16,16,16,16,16,15,15,15,16,16,16,16,16,16,15,16,16,16,16,16,16,16,16,16,15,16,16,16,16,15]

## Perguntas 6-8
A seguinte lista de 7 listas numéricas tem as <u>idades de todos os alunos de 8º</u> ano do AEFGA, <u>organizadas turma a turma</u>.

<code><b>[</b> [13,13,13,13,13,13,13,12,13,13,13,13,13,13,13,13,13,13,13,13,12,13], [13,12,13,13,13,13,16,13,13,14,12,13,13,13,14,13,12,13,17,13,12,13,13,14,13,17,12], [12,13,13,12,13,12,13,13,13,13,13,13,13,14,13,13,14,13,13,12,12,13,12,13,12,13,13,13,13], [13,13,13,13,14,13,13,13,13,13,14,13,14,12,13,13,13,13,13,12,13,13], [16,12,13,13,12,13,14,13,13,13,13,13,13,13,13,13,13,13,14,13,13,13,14,14], [13,13,13,13,13,13,13,13,12,13,12,13,13,13,13,13,12,15,14,12,13,13,13,13,13,13,12,13], [13,13,12,14,13,13,13,13,13,13,14,12,12,13,13,13,13,13,14,13,13,13] <b>]</b></code>

### Pergunta 6
Escreve e corre uma instrução para <u>dar o identificador `idades8`</u> a esta lista.
<br> Depois corre a seguinte instrução:

`idades8 |> map maximum |> maximum`

Qual foi o resultado dado pelo Jupyter?

**Solução:**

In [8]:
idades8 = [ [13,13,13,13,13,13,13,12,13,13,13,13,13,13,13,13,13,13,13,13,12,13],[13,12,13,13,13,13,16,13,13,14,12,13,13,13,14,13,12,13,17,13,12,13,13,14,13,17,12],[12,13,13,12,13,12,13,13,13,13,13,13,13,14,13,13,14,13,13,12,12,13,12,13,12,13,13,13,13],[13,13,13,13,14,13,13,13,13,13,14,13,14,12,13,13,13,13,13,12,13,13],[16,12,13,13,12,13,14,13,13,13,13,13,13,13,13,13,13,13,14,13,13,13,14,14],[13,13,13,13,13,13,13,13,12,13,12,13,13,13,13,13,12,15,14,12,13,13,13,13,13,13,12,13],[13,13,12,14,13,13,13,13,13,13,14,12,12,13,13,13,13,13,14,13,13,13] ]

In [9]:
idades8 |> map maximum |> maximum

17

### Pergunta 7
O resultado da pergunta anterior podia ser utilizado para responder a qual destas perguntas?

⬜ Quantas turmas de 8º ano existem?
<br>⬜ Quantos alunos do 8º ano têm 17 anos?
<br>⬜ Quem é o aluno mais velho do 8º ano?
<br>⬜ Quantos anos tem o aluno mais velho do 8º ano?

**Solução:**

A lista chamada **`idades8` é uma lista de listas** numéricas, onde cada lista numérica contém as idades de uma certa turma.

<code>idades8 = <b>[</b> [13,13,13,13,13,13,13,12,13,13,13,13,13,13,13,13,13,13,13,13,12,13]<b>,</b> [13,12,13,13,13,13,16,13,13,14,12,13,13,13,14,13,12,13,17,13,12,13,13,14,13,17,12]<b>,</b> [12,13,13,12,13,12,13,13,13,13,13,13,13,14,13,13,14,13,13,12,12,13,12,13,12,13,13,13,13]<b>,</b> [13,13,13,13,14,13,13,13,13,13,14,13,14,12,13,13,13,13,13,12,13,13]<b>,</b> [16,12,13,13,12,13,14,13,13,13,13,13,13,13,13,13,13,13,14,13,13,13,14,14]<b>,</b> [13,13,13,13,13,13,13,13,12,13,12,13,13,13,13,13,12,15,14,12,13,13,13,13,13,13,12,13]<b>,</b> [13,13,12,14,13,13,13,13,13,13,14,12,12,13,13,13,13,13,14,13,13,13] <b>]</b></code>

A instrução `idades8 |> map maximum |> maximum` faz duas transformações.
<br>Quando escrevemos `idades8 |> map maximum`, estamos a pedir ao Jupyter que pegue em cada lista numérica e lhe faça a transformação `maximum`:

In [10]:
idades8 |> map maximum

[13,17,14,14,16,15,14]

O resultado é uma lista que tem **a maior idade de cada turma**.
<br>Se, em seguida, fizermos a transformação `maximum`, vamos ficar com **a maior das maiores** idades:

In [11]:
idades8 |> map maximum |> maximum

17

Isto é, nós descobrimos a maior idade de entre as idades dos alunos de todas turmas de 8º ano.

⬜ Quantas turmas de 8º ano existem?
<br>⬜ Quantos alunos do 8º ano têm 17 anos?
<br>⬜ Quem é o aluno mais velho do 8º ano?
<br>✅ **Quantos anos tem o aluno mais velho do 8º ano?**

### Pergunta 8
<u>Quantos</u> alunos de 8º ano há no AEFGA?

<small><b>Pista:</b> Começa por contar quantos alunos há em cada turma.</small>

**Solução:**

Como sugerido na pista, vamos começar por contar quantos alunos há em cada turma.

<code>idades8 = <b>[</b> [13,13,13,13,13,13,13,12,13,13,13,13,13,13,13,13,13,13,13,13,12,13]<b>,</b> [13,12,13,13,13,13,16,13,13,14,12,13,13,13,14,13,12,13,17,13,12,13,13,14,13,17,12]<b>,</b> [12,13,13,12,13,12,13,13,13,13,13,13,13,14,13,13,14,13,13,12,12,13,12,13,12,13,13,13,13]<b>,</b> [13,13,13,13,14,13,13,13,13,13,14,13,14,12,13,13,13,13,13,12,13,13]<b>,</b> [16,12,13,13,12,13,14,13,13,13,13,13,13,13,13,13,13,13,14,13,13,13,14,14]<b>,</b> [13,13,13,13,13,13,13,13,12,13,12,13,13,13,13,13,12,15,14,12,13,13,13,13,13,13,12,13]<b>,</b> [13,13,12,14,13,13,13,13,13,13,14,12,12,13,13,13,13,13,14,13,13,13] <b>]</b></code>

Cada lista numérica dentro da lista `idades8` tem as idades dos alunos de uma certa turma. Se soubermos quantas idades há em cada turma, ficamos a saber quantos alunos tem cada turma.
<br>Então começamos por pedir ao Jupyter para **contar** quantos números há **em cada** lista:

In [12]:
idades8 |> map length

[22,27,29,22,24,28,22]

Depois basta somarmos as contagens de cada turma, para ficarmos a saber o total de alunos:

In [13]:
idades8 |> map length |> sum

174

### Pergunta 9
Observa o resultado de cada instrução e completa-as com as transformações
- `reverse`,
- `sort`,
- `words |> map reverse |> unwords` ou
- `words |> map sort |> unwords`.

Instrução: `aifoi = "Frei Goncalo de Azevedo foi um cavaleiro do seculo XVI"`

Instrução: `aifoi |> ______________________`
<br>Resultado: `"IVX oluces od orielavac mu iof odevezA ed olacnoG ierF"`

Instrução: `aifoi |> ______________________`
<br>Resultado: `"Feir Gaclnoo de Adeeovz fio mu aaceilorv do celosu IVX"`

Instrução: `aifoi |> ______________________`
<br>Resultado: `"ierF olacnoG ed odevezA iof mu orielavac od oluces IVX"`

Instrução: `aifoi |> ______________________`
<br>Resultado: `"         AFGIVXaaacccdddeeeeeefiiilllmnooooooorrsuuvvz"`

**Solução:**

A 1ª instrução serve só para dar o identificador (nome) `aifoi` à _string_ `"IVX oluces od orielavac mu iof odevezA ed olacnoG ierF"`.

In [14]:
aifoi = "Frei Goncalo de Azevedo foi um cavaleiro do seculo XVI"

---
**1º Resultado: `"IVX oluces od orielavac mu iof odevezA ed olacnoG ierF"`**

Observando o resultado, vemos que os **caracteres** da _string_ `aifoi` foram trocados **do fim para o princípio**. Vemos que o primeiro caractere da `aifoi` (o `'F'`) passou a ser o último, o segundo caractere (`'r'`) passou para penúltimo, ..., o penúltimo caractere (`'V'`) passou para segundo, e o último caratere (o `'I'`) passou a ser o primeiro da _string_.

Para o resultado ser este, a _string_ teve de ser invertida usando a transformação `reverse`:

In [15]:
aifoi |> reverse

"IVX oluces od orielavac mu iof odevezA ed olacnoG ierF"

---
**2º Resultado: `"Feir Gaclnoo de Adeeovz fio mu aaceilorv do celosu IVX"`**

Observando este resultado, vemos que cada palavra continua no mesmo sítio, mas que as letras de cada palavra foram ordenadas alfabeticamente. Portanto, primeiro pedimos ao Jupyter para **partir a frase numa lista de palavras** (`words`), depois pedimos-lhe para **ordenar** as letras de **cada** palavra da lista (`map sort`), e depois para pegar nessa lista de palavras estranhas e **juntá-las numa frase** (`unwords`):

In [16]:
aifoi |> words

["Frei","Goncalo","de","Azevedo","foi","um","cavaleiro","do","seculo","XVI"]

In [17]:
aifoi |> words |> map sort

["Feir","Gaclnoo","de","Adeeovz","fio","mu","aaceilorv","do","celosu","IVX"]

In [18]:
aifoi |> words |> map sort |> unwords

"Feir Gaclnoo de Adeeovz fio mu aaceilorv do celosu IVX"

---
**3º Resultado: `"ierF olacnoG ed odevezA iof mu orielavac od oluces IVX"`**

Observando este resultado, vemos que cada palavra continua no mesmo sítio, mas que as letras de cada palavra foram invertidas. Portanto, primeiro pedimos ao Jupyter para **partir a frase numa lista de palavras** (`words`), depois pedimos-lhe para **inverter** as letras de **cada** palavra da lista (`map reverse`), e depois para pegar nessa lista de palavras estranhas e **juntá-las numa frase** (`unwords`):

In [19]:
aifoi |> words

["Frei","Goncalo","de","Azevedo","foi","um","cavaleiro","do","seculo","XVI"]

In [20]:
aifoi |> words |> map reverse

["ierF","olacnoG","ed","odevezA","iof","mu","orielavac","od","oluces","IVX"]

In [21]:
aifoi |> words |> map reverse |> unwords

"ierF olacnoG ed odevezA iof mu orielavac od oluces IVX"

---
**4º Resultado: `"         AFGIVXaaacccdddeeeeeefiiilllmnooooooorrsuuvvz"`**

Observando este resultado, vemos que os **caracteres** da _string_ `aifoi` foram **ordenados**. Ficaram em primeiro <u>os 9 espaços</u> da frase, depois <u>as 6 maiúsculas</u> por ordem alfabética (`'A'`, `'F'`, `'G'`, `'I'`, `'V'` e `'X'`), e depois <u>as minúsculas</u> por ordem alfabética (a ordem pela qual este caracteres surgem na tabela de ASCII).

Para o resultado ser este, a _string_ teve de ser ordenada usando a transformação `sort`:

In [22]:
aifoi |> sort

"         AFGIVXaaacccdddeeeeeefiiilllmnooooooorrsuuvvz"

## Desafio Extra

- Calcula a soma das idades dos alunos do 6º ano.
- Calcula a soma das idades dos alunos do 8º ano.

<u>Qual a soma maior?</u>

**Solução:**

Recordemos as duas listas:

`idades6 = [11,10,11,11,11,10,11,11,11,11,11,11,11,11,11,11,11,10,11,11,11,10,11,11,11,11,10,11,12,11,11,12,11,11,11,11,11,11,10,11,10,11,11,11,11,11,11,11,10,10,11,11,11,12,11,12,11,11,11,10,11,12,11,10,11,14,11,11,10,11,11,13,11,10,11,11,11,11,11,11,10,11,10,11,11,11,12,11,10,11,11,11,11,11,11,11,10,11,11,12,11,12,11,11,11,11,11,10,11,11,11,10,11,10,10,11,10,10,10,10,11,10,10,11,11,12,11,11,11,11,11,11,10,10,10,11,11,11,11,11,11,10,11,11,11,11,11,11,11,11,11,10,11,11,11,11,10]`

<code>idades8 = <b>[</b> [13,13,13,13,13,13,13,12,13,13,13,13,13,13,13,13,13,13,13,13,12,13]<b>,</b> [13,12,13,13,13,13,16,13,13,14,12,13,13,13,14,13,12,13,17,13,12,13,13,14,13,17,12]<b>,</b> [12,13,13,12,13,12,13,13,13,13,13,13,13,14,13,13,14,13,13,12,12,13,12,13,12,13,13,13,13]<b>,</b> [13,13,13,13,14,13,13,13,13,13,14,13,14,12,13,13,13,13,13,12,13,13]<b>,</b> [16,12,13,13,12,13,14,13,13,13,13,13,13,13,13,13,13,13,14,13,13,13,14,14]<b>,</b> [13,13,13,13,13,13,13,13,12,13,12,13,13,13,13,13,12,15,14,12,13,13,13,13,13,13,12,13]<b>,</b> [13,13,12,14,13,13,13,13,13,13,14,12,12,13,13,13,13,13,14,13,13,13] <b>]</b></code>

Como a lista **`idades6` é uma lista de números**, para somar todos os seus números basta usar a transformação `sum`:

In [23]:
idades6 |> sum

1708

Mas a lista **`idades8` é uma lista de listas numéricas**! Não é possível somar diretamente todas as idades.
<br>Temos de começar por somar os números de cada lista numérica:

In [24]:
idades8 |> map sum

[284,360,372,287,317,362,286]

Ficamos com uma lista numérica com as somas de idades de cada turma. Agora, podemos pegar nessa lista e somar os seus números:

In [25]:
idades8 |> map sum |> sum

2268

**Resposta:** A soma das idades dos alunos do 8º ano (2268) é maior do que a do 6º ano (1708).